File: week4/stress_test/binary_search.cpp
#include <iostream>
#include <cassert>
#include <vector>
#include <random>
#include <bits/stdc++.h>

using std::vector;
using std::sort;

int binary_search(vector<int> &a, int x)
{
int left = 0, right = (int)a.size(); 

  while (left <= right){
    int mid = left + (int)(right-left)/2;

    if (x == a[mid]){
      while (a[mid - 1] == x){
        if (mid != 0)
          mid -= 1;
      }
      return mid;
    }
    else if (x < a[mid])
      right = mid-1;
    else
      left = mid+1;
  }
  return -1;
}

int linear_search(const vector<int> &a, int x) {
  for (size_t i = 0; i < a.size(); ++i) {
    if (a[i] == x) return i;
  }
  return -1;
}

int main() {
  while (true){
    int n = 1 + rand() % 10000;
    vector<int> a(n);
    for (int i=0; i<n; i++){
      a[i] = rand() % 10000;
    }
    sort(a.begin(), a.end());
    int x = rand() % 10000;
    int linear_out = linear_search(a, x);
    int binary_out = binary_search(a, x);
    if (linear_out == binary_out){
      std::cout<< "OK!!" << "\n";
    }
    else{
      std::cout << "Array : ";
      for (int i=0; i<n; i++){
        std::cout << a[i] << " ";
      }

      std::cout<< "\n";
      std::cout<< "element : " << x << "\n";

      std::cout << "linear_output : " << linear_out << "\n";      
      std::cout << "binary_output : " << binary_out << "\n";
      break;
    }
  }
}

File: week4/stress_test/binary_search.py
# Uses python3
from numpy import random

def binary_search(a, x):
    left, right = 0, len(a)-1
    # write your code here
    while left <= right:
        mid = left + (right - left)//2
        if a[mid] == x:
            while a[mid-1] == x:
                if mid == 0:
                    break
                else:
                    mid -= 1

            return mid
        elif a[mid] > x:
            right = mid  - 1
        else:
            left = mid + 1
    return -1

def linear_search(a, x):
    for i in range(len(a)):
        if a[i] == x:
            return i
    return -1

if __name__ == '__main__':

    while True:
        n = random.randint(1, 10)
        a = []
        for i in range(n):
            a.append(random.randint(1, 10))

        a.sort()
        x = random.randint(1, 10)

        res_linear = linear_search(a, x)
        res_binary = binary_search(a, x)

        if res_binary != res_linear:
            print('Array : ',a)
            print('Key : ',x)
            print('binary_search : ',res_binary)
            print('linear_search : ',res_linear)
            break;
        else:
            print('OK!!')